File: python/day1.py

def getInput():
    data = []
    with open("../inputs/day1.txt", "r") as file:
        for line in file:
            data.append([int(x) for x in line.strip("\n").split("   ")])
    return data

def splitIntoArrays(data):
    arrL, arrR = [], []
    for pair in data:
        arrL.append(pair[0])
        arrR.append(pair[1])

    return [arrL, arrR]

def solutionA(data):
    arrL, arrR = splitIntoArrays(data)

    arrL.sort()
    arrR.sort()

    sum = 0
    for i in range(len(arrL)):
        sum += abs(arrL[i] - arrR[i])

    return sum

def countElements(arr):
    counter = dict()
    for ele in arr:
        if ele in counter:
            counter[ele] += 1
        else:
            counter[ele] = 1

    return counter

def solutionB(data):
    arrL, arrR = splitIntoArrays(data)

    counterL = countElements(arrL)
    counterR = countElements(arrR)

    sum = 0
    for key in counterL:
        if key in counterR:
            sum += counterR[key] * key

    return sum

data = getInput()
print(solutionA(data))
print(solutionB(data))

File: typescript/src/day1.ts
import fs from 'fs';

function getInput(): Array<string>{
    const data = fs.readFileSync('../../inputs/day1.txt', 'utf8');
    const lines = data.split('\n')

    return lines;
}

function processInput(input: Array<string>): Array<Array<number>>{
    const locations:Array<Array<number>> = [[], []];
    for(let line of input){
        const lineLocations = line.split('   ').map(Number);
        locations[0].push(lineLocations[0]);
        locations[1].push(lineLocations[1]);
    }

    return locations;
}

function sortLocations(locations: Array<Array<number>>): Array<Array<number>>{
    locations[0].sort((a, b) => a - b);
    locations[1].sort((a, b) => a - b);

    return locations;
}

function calculateDifference(locations1: Array<number>, location2: Array<number>): number{
    let total = 0;
    for(let i=0; i<locations1.length; i++){
        total += Math.abs(locations1[i] - location2[i]);
    }

    return total;
}

function countLocations(locations: Array<number>): {[key: number]: number}{
    const counts: {[key: number]: number} = {};
    for(let location of locations){
        if(counts[location]){
            counts[location]++;
        }else{
            counts[location] = 1;
        }
    }

    return counts;
}

function calculateSimilarity(location1: {[key: number]: number}, location2: {[key: number]: number}): number{

    let similarity = 0;
    for(let key in location1){
        if(location2[key]){
            similarity += Number(key) * location2[key];
        }
    }

    return similarity;
}

function solutionA(input:Array<string>):void{
    const locations = processInput(input);
    const sortedLocations = sortLocations(locations);
    const difference = calculateDifference(sortedLocations[0], sortedLocations[1]);
    console.log(difference);
}

function solutionB(input:Array<string>):void{
    const locations = processInput(input);
    const countsLocation1 = countLocations(locations[0]);
    const countsLocation2 = countLocations(locations[1]);
    const similarity = calculateSimilarity(countsLocation1, countsLocation2);
    console.log(similarity);
}


function main(){
    const input = getInput();
    solutionA(input)
    solutionB(input)
}

main();